### DISCUSSION TITLE
Programme aus Homematic in Home Assistant

### DISCUSSION THREAD

**Discussion by riesi77** [SukramJ/aiohomematic#559] [Q&A]
Hallo,
möglicherweise übersehe ich etwas - aber für mich wäre es die einfachste Möglichkeit, die Programme aus Homematic als Entities zu haben. Gibt es Pläne, das bereitzustellen? Wie wäre es sonst möglich, Homematic Programme aus Home Assistant aufzurufen?

**Comment by danielperna84** [comment]
Hier mögen die Meinungen vielleicht auseinander gehen. Aber ich bin der Ansicht, dass HA keine Programme auf der CCU ausführen sollte. Entweder funktionieren die Programme auf der CCU autark, man nutzt Direktverknüpfungen, oder die Automation läuft komplett in HA. So weiß man bei einem Fehlerfall dann auch direkt wo man nach dem Problem suchen muss.

**Reply by riesi77** [reply]
Die Möglichkeit würde ja weiter bestehen - also die Logik komplett in HM zu belassen. Nur hätte man die Möglichkeit und Flexibilität, sie aus HA nur aufzurufen und nicht nachprogrammieren zu müssen.

**Reply by danielperna84** [reply]
Ich hätte vielleicht noch dazu sagen sollen: das meiste läuft über XML-RPC. Und dort hat eq3 keine Möglichkeit eingebaut mit den Programmen auf der CCU zu interagieren. Vielleicht geht das mit dem Rega-Script-Kram der auch für den Cache usw. benutzt wird. Aber von dem Thema habe ich nicht wirklich Ahnung.

@Baxxy13 weißt du ob man per Rega mit den Programmen auf der CCU arbeiten kann?

**Reply by SukramJ** [reply]
In der Json-API gibt es ein `Program.execute` und `Program.getAll`. Hab ich aber noch nicht verwendet.

**Reply by riesi77** [reply]
Ich weiß nicht, ob euch das hilft. Aber es gibt auf http ebene einen Aufruf <ip>/config/xmlapi/programlist.cgi, das mit die Programme und deren ID liefert. Und über <ip>config/xmlapi/runprogram.cgi?program_id=<program id> lasse ich das Programm laufen. Ich mache das z.b., um über ein Browser Bookmark/ Home Screen Widget meine Kaffeemaschine einzuschalten, die über eine HM-IP steuerbare Steckdose eingesteckt ist.

**Reply by SukramJ** [reply]
So richtig spricht auch nichts dagegen ein CCU Program auf Knopfdruck zu starten.

**Reply by SukramJ** [reply]
Mit Version 1.13.1 werden die CCU Programme (keine internen) als `button` (initial deaktiviert) unterhalb der Service Entität angezeigt.

**Reply by riesi77** [reply]
Danke, das ist echt genial. Wäre es auch möglich, die Programme als Skript anzuzeigen? Diese wäre z.B. über Shortcuts besser anzusteuern.

**Reply by SukramJ** [reply]
Nein, das geht nicht. Was stimmt mit den Button nicht?

**Reply by ThomDietrich** [reply]
Ich muss hier Daniel voll zustimmen. Bevor du versuchst Programme aus Home Assistant heraus auszuführen oder gar in breitere Logik in Form von Skripten integrieren zu wollen, implementiere deine Logik besser direkt und vollständig in Home Assistant. Du wirst sehen, dass das schnell deutliche Vorteile bringt.

Schau auch mal hier vorbei: https://github.com/danielperna84/hahomematic/discussions/475

**Reply by riesi77** [reply]
Konkreter Hintergrund ist, dass das Dienstaufruf Widget der Home Assistant App unter Android nur Dienste - also z.b. Script aufrufen kann, aber keine Buttons. Ich habe aber natürlich die Möglichkeit, entweder den Button in ein Script zu wrappen, oder - wie von euch bereits mehrfach angemerkt, die gesamte Logik in Home Assistant zu überführen.
Herzlichen Dank nochmal. Gibt es eine Möglichkeit, an Euer Projekt zu spenden?

**Reply by SukramJ** [reply]
Der Button hat doch auch einen Dienst. `button.press`

**Reply by riesi77** [reply]
> Der Button hat doch auch einen Dienst. `button.press`

Aaah. Da erkennt man dann doch, dass ich noch neu in HA bin. Danke. Damit funktioniert's.

**Comment by kwithus** [comment]
Im letzten Update steht etwas von Programme als Buttons, was genau kann man darunter verstehen? Ich konnte jedenfalls bei mir keine entsprechenden Buttons in den entities finden, die nach einem entsprechenden Prog. auf der CCU Klingen

**Reply by Baxxy13** [reply]
Man muss die erst aktivieren. 
Oder es besser lassen und seine Logiken direkt in HA umsetzen. ;-)
![HA-Programmbuttons](https://user-images.githubusercontent.com/54898249/183311347-c2d3fd96-6522-4100-ba95-fa09d7c57eef.JPG)

**Comment by Baxxy13** [comment]
Ich bin hier mal voll bei Daniel und finde auch das die WebUI-Programme für sich (durch ihre Triggerbedingungen) arbeiten sollten.
Ein "externes anstubsen" ist m.E. schlechter Programmierstil. 

Die Begründung ist einfach...
`Program.execute` (was vermutlich genutzt wird) macht nichts anderes als alle Aktionen im 1.DANN des WebUI-Programmes auszuführen. 
Völlig ohne Bedingungsprüfung, also egal ob die WENN-Bedingungen "WAHR" ergeben oder nicht. 
Da kann man doch gleich in HA bei Event X Aktion Y auslösen, warum der Umweg über das quasi nutzlose WebUI-Programm? 

Stören tun mich die deaktivierten "Programm-Buttons" nicht, einen Nutzen sehe ich in den Buttons aber auch nicht,
Zumindest fällt mir kein Szenario ein wo ich sowas einsetzen würde, aber vielleicht bringt ja jemand mal ein gutes Beispiel.

**Reply by danielperna84** [reply]
Also ich muss zugeben, dass das Argument mit den Neueinsteigern die auf der CCU schon viele produktive Programme haben schon nicht ganz verkehrt ist. So kann man Stück für Stück migrieren, und ggf. auch wieder zurück rudern wenn man merkt, dass man HA doch nicht so mag.

Und ein Vorteil wäre, dass man über diesen Umweg Cuxd Geräte (blind) steuern kann. Das mag dem ein oder anderen vielleicht lieber sein als diese Geräte gar nicht nutzen zu können.

**Reply by Baxxy13** [reply]
Das mit den CUxD Geräten ist ein interessanter Ansatz, das lasse ich mal gelten. ;-)

Bei den vielen Programmen wüsste ich nicht wie da ein manuelles anstupsen helfen sollte Schritt für Schritt nach HA zu migrieren. Wie geschrieben wird die Programmlogik dabei völlig außer Kraft gesetzt. 
Ich würde als Fallback die WebUI-Programme einfach deaktivieren.
